test/python/transpiler/test_kak_over_optimization.py

```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test KAK over optimization"""
+
+import unittest
+import numpy as np
+from qiskit import QuantumCircuit, QuantumRegister, transpile
+from qiskit.test import QiskitTestCase
+
+
+class TestKAKOverOptim(QiskitTestCase):
+    """ Tests to verify that KAK decomposition
+    does not over optimize.
+    """
+
+    def test_cz_optimization(self):
+        """ Test that KAK does not run on a cz gate """
+        qr = QuantumRegister(2)
+        qc = QuantumCircuit(qr)
+
+        qc.cz(qr[0], qr[1])
+
+        cz_circ = transpile(qc, None, coupling_map=[[0, 1], [1, 0]],
+                            basis_gates=['u1', 'u2', 'u3', 'id', 'cx'],
+                            optimization_level=3)
+        ops = cz_circ.count_ops()
+        self.assertEqual(ops['u2'], 2)
+        self.assertEqual(ops['cx'], 1)
+        self.assertFalse('u3' in ops.keys())
+
+    def test_cu1_optimization(self):
+        """ Test that KAK does run on a cu1 gate and
+        reduces the cx count from two to one.
```

Comment:
I ran this test locally without your patch applied and was not able to reproduce a case with 2 cx gates in the output ops. It always only had 1 cx listed.